homeassistant/components/sia/hub.py

```diff
@@ -0,0 +1,197 @@
+    CONF_ZONE,
+    DEVICE_CLASS_TIMESTAMP,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import Event, HomeAssistant
+from homeassistant.helpers import device_registry as dr
+from homeassistant.util.dt import utcnow
+from homeassistant.util.json import load_json
+
+from .alarm_control_panel import SIAAlarmControlPanel
+from .const import (
+    CONF_ACCOUNT,
+    CONF_ACCOUNTS,
+    CONF_ENCRYPTION_KEY,
+    CONF_PING_INTERVAL,
+    CONF_ZONES,
+    DEVICE_CLASS_ALARM,
+    DOMAIN,
+    HUB_SENSOR_NAME,
+    LAST_MESSAGE,
+    UTCNOW,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SIAHub:
+    """Class for SIA Hubs."""
+
+    def __init__(
+        self, hass: HomeAssistant, hub_config: dict, entry_id: str, title: str
+    ):
+        """Create the SIAHub."""
+        self._hass = hass
+        self.states = {}
+        self._port = int(hub_config[CONF_PORT])
+        self.entry_id = entry_id
+        self._title = title
+        self._accounts = hub_config[CONF_ACCOUNTS]
+        self.shutdown_remove_listener = None
+        self._reactions = None
+        self._zones = [
+            {
+                CONF_ACCOUNT: a[CONF_ACCOUNT],
+                CONF_ZONE: z,
+                CONF_SENSORS: [DEVICE_CLASS_ALARM],
+            }
+            for a in self._accounts
+            for z in range(1, int(a[CONF_ZONES]) + 1)
+        ]
+        self.sia_accounts = [
+            SIAAccount(a[CONF_ACCOUNT], a.get(CONF_ENCRYPTION_KEY))
+            for a in self._accounts
+        ]
+        self.sia_client = SIAClient(
+            "", self._port, self.sia_accounts, self.update_states
+        )
+        self._create_sensors()
+
+    async def async_setup_hub(self):
+        """Add a device to the device_registry, register shutdown listener, load reactions."""
+        device_registry = await dr.async_get_registry(self._hass)
+        port = self._port
+        for acc in self._accounts:
+            account = acc[CONF_ACCOUNT]
+            device_registry.async_get_or_create(
+                config_entry_id=self.entry_id,
+                identifiers={(DOMAIN, port, account)},
+                name=f"{port} - {account}",
+            )
+        self.shutdown_remove_listener = self._hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, self.async_shutdown
+        )
+        self._reactions = load_json("homeassistant/components/sia/reactions.json")
+
+    async def async_shutdown(self, _: Event):
+        """Shutdown the SIA server."""
+        await self.sia_client.stop()
+
+    def _create_sensors(self):
+        """Create all the sensors."""
+        for zone in self._zones:
+            ping = self._get_ping_interval(zone[CONF_ACCOUNT])
+            for entity_type in zone[CONF_SENSORS]:
+                entity_id, entity_name = self._get_entity_id_and_name(
+                    zone[CONF_ACCOUNT], zone[CONF_ZONE], entity_type
+                )
+                self.states[entity_id] = SIAAlarmControlPanel(
```

Comment:
I think it's better to create something else here (maybe a dict), and the actual `SIAAlarmControlPanel` instances in `async_setup_entry` for alarm control panel.

Reply:
I'm not sure if that will work since all the update logic happens inside the hub as well, but I'll check.

Reply:
I'm still not a big fan of this. It seems that you can store the parameters to `SIAAlarmControlPanel` here and initialize it in `alarm_control_panel.py`.